Gray failure allows storage servers to complain

spraza opened the pull request:
# Description

There are two main changes:

1. This PR lets storage servers, both in primary and/or remote DCs, to complain about workers in the txn subsystem. For example, if a remote tlog is degraded (e.g. network issue), remote storage servers to which this tlog is a buddy of, can report the tlog to CC. This gives an opportunity for CC and gray failure to then exclude this degraded tlog and trigger recovery. This is especially useful in pathological network failure modes e.g. tlog's network link to SS is degraded, but to log router is fine. In this scenario, log routers won't complain about this tlog, and this tlog will remain degraded and exist in the cluster. This PR fixes that problem.

2. The existing remote tlog simulation test is improved in different ways. The test now _always_ checks that it passes through a series of test phases aka `expectedStatePaths`. Previously, the test path was conditionally (based on process placement) expected to end in `SS_LAG_NORMAL`. Now, that condition is removed, and we unconditionally do the check. To make this unconditional, a new test phase `CLOGGED_REMOTE_TLOG_EXCLUDED` is added, and is computed via server dbInfo. This state is always expected to be reached even if `SS_LAG_NORMAL` is not reached. Finally, there could be a lot of test state paths. In order to not hard code all possible state paths, a concept of prefix matching is introduced, and every path in `expectedStatePaths` can be marked as prefix true or false, depending on how strict we want the matching to be (at the end in `check()`). 

# Testing

- Enhanced current simulation test to clog only tlog->SS network if the feature flag in this PR is on. This means the test can give signal for general network problems, but also specifically for tlog->SS network problems.
- 100K Joshua: `20241104-234909-praza-5bbe014302e5c6b1350ad9a294fbc7f0a65397 compressed=True data_size=36041745 duration=5521908 ended=100000 fail=2 fail_fast=10 max_runs=100000 pass=99998 priority=100 remaining=0 runtime=0:52:26 sanity=False started=100000 stopped=20241105-004135 submitted=20241104-234909 timeout=5400 username=praza-5bbe014302e5c6b1350ad9a294fbc7f0a653975d`. 2 failures but don't seem related to my change, will confirm. 

# Code-Reviewer Section

The general pull request guidelines can be found [here](https://github.com/apple/foundationdb/wiki/FoundationDB-Commit-Process).

Please check each of the following things and check *all* boxes before accepting a PR.

- [ ] The PR has a description, explaining both the problem and the solution.
- [ ] The description mentions which forms of testing were done and the testing seems reasonable.
- [ ] Every function/class/actor that was touched is reasonably well documented.

## For Release-Branches

If this PR is made against a release-branch, please also check the following:

- [ ] This change/bugfix is a cherry-pick from the next younger branch (younger `release-branch` or `main` if this is the youngest branch)
- [ ] There is a good reason why this PR needs to go into a release branch and this reason is documented (either in the description above or in a linked GitHub issue)

jzhou77 commented on fdbclient/ServerKnobs.cpp:
```diff
@@ -1101,7 +1101,9 @@ void ServerKnobs::initialize(Randomize randomize, ClientKnobs* clientKnobs, IsSi
 	init( WORKER_HEALTH_REPORT_RECENT_DESTROYED_PEER,           true );
-	init( GRAY_FAILURE_ENABLE_TLOG_RECOVERY_MONITORING,         true );
+	init( GRAY_FAILURE_ENABLE_TLOG_RECOVERY_MONITORING,         true );	
+	init( GRAY_FAILURE_ALLOW_PRIMARY_SS_TO_COMPLAIN,           false ); if (isSimulated && deterministicRandom()->coinflip()) GRAY_FAILURE_ALLOW_PRIMARY_SS_TO_COMPLAIN = true;
```
Nit: can be simplified as `if (isSimulated) GRAY_FAILURE_ALLOW_PRIMARY_SS_TO_COMPLAIN = deterministicRandom()->coinflip();`

spraza replied:
Resolved in #11757. Also updated other such instances.

jzhou77 commented on fdbserver/worker.actor.cpp:
```diff
@@ -1356,20 +1392,77 @@ UpdateWorkerHealthRequest doPeerHealthCheck(const WorkerInterface& interf,
+		    wait(NativeAPI::getServerListAndProcessClasses(&tr));
+		PrimaryAndRemoteAddresses storageServers;
+		for (auto& [ssi, _] : results) {
+			const auto primaryDCId = getPrimaryDCId(dbInfo->get());
```
move before entering this loop

spraza replied:
Good catch! Fixed in #11757.

jzhou77 commented on fdbserver/worker.actor.cpp:
```diff
@@ -1356,20 +1392,77 @@ UpdateWorkerHealthRequest doPeerHealthCheck(const WorkerInterface& interf,
-			req = doPeerHealthCheck(interf, locality, dbInfo, req, enablePrimaryTxnSystemHealthCheck);
+			state Optional<PrimaryAndRemoteAddresses> storageServers;
+			if (db.present()) {
+				Optional<PrimaryAndRemoteAddresses> storageServers_ = wait(getStorageServers(db.get(), dbInfo));
```
can be simplified as `store(storageServers, getStorageServers(db.get(), dbInfo));

spraza replied:
Reads better, thanks for telling about this utility. Fixed in #11757. Since `store` returns `Future`, also added `wait` to your code suggestion.

jzhou77 reviewed (approved):
LGTM

spraza commented:
> @jzhou77 Merged but just realized there was some minor feedback, sorry didn't see it. Will send a separate PR.

Addressed all feedback in #11757.
